cachi2/core/package_managers/npm.py

```diff
@@ -173,6 +185,199 @@ def get_sbom_components(self) -> list[dict[str, str]]:
+        }
+
+
+def _update_vcs_url_with_full_hostname(vcs: str) -> str:
+    """Update VCS URL with full hostname.
+
+    Transform github:kevva/is-positive#97edff6
+    into git+ssh://github.com/kevva/is-positive.git#97edff6
+
+    :param vcs: VCS URL to be modified with full hostname and file extension
+    :return: Updated VCS URL
+    """
+    host, _, path = vcs.partition(":")
+    namespace_repo, _, ref = path.partition("#")
+    suffix_domain = "org" if host == "bitbucket" else "com"
+
+    vcs = f"git+ssh://{host}.{suffix_domain}/{namespace_repo}.git"
+    if ref:
+        vcs = f"{vcs}#{ref}"
+    return vcs
+
+
+def _extract_git_info_npm(vcs_url: str) -> Dict[str, str]:
+    """
+    Extract important info from a VCS requirement URL.
+
+    Given a URL such as git+ssh://user@host/namespace/repo.git#9e164b970
+
+    this function will extract:
+    - the "clean" URL: ssh://user@host/namespace/repo.git
+    - the git ref: 9e164b970
+
+    The clean URL and ref can be passed straight to scm.Git to fetch the repo.
+    The host, namespace and repo will be used to construct the file path under deps/npm.
+
+    :param vcs_url: The URL of a VCS requirement, must be valid (have git ref in path)
+    :return: Dict with url, ref, host, namespace and repo keys
+    """
+    # If scheme is git+protocol://, keep only protocol://
+    # Do this before parsing URL, otherwise urllib may not extract URL params
+    if vcs_url.startswith("git+"):
+        vcs_url = vcs_url[len("git+") :]
+
+    clean_url, _, ref = vcs_url.partition("#")
+
+    url = urlparse(clean_url)
+    namespace_repo = url.path.strip("/")
+    if namespace_repo.endswith(".git"):
+        namespace_repo = namespace_repo[: -len(".git")]
+
+    # Everything up to the last '/' is namespace, the rest is repo
+    namespace, _, repo = namespace_repo.partition("/")
+
+    vcs_url_info = {
+        "url": clean_url,
+        "ref": ref.lower(),
+        "namespace": namespace,
+        "repo": repo,
+    }
+
+    for key, value in vcs_url_info.items():
+        if not value:
+            raise UnexpectedFormat(f"{vcs_url} is not valid VCS url. {key} is missing.")
+
+    if url.hostname:
+        vcs_url_info["host"] = url.hostname
+    else:
+        raise UnexpectedFormat(f"{vcs_url} is not valid VCS url. Host is missing.")
+
+    return vcs_url_info
+
+
+def _clone_repo_pack_archive(
+    vcs: str,
+    download_dir: RootedPath,
+) -> RootedPath:
+    """
+    Clone a repository and pack its content as tar.
+
+    :param url: URL for file download
+    :param download_dir: Output folder where dependencies will be downloaded
+    :raise FetchError: If download failed
+    """
+    info = _extract_git_info_npm(vcs)
+    download_path = RootedPath(
+        Path(
+            download_dir,
+            info["host"],  # host
+            info["namespace"],
+            info["repo"],
+            f'{info["repo"]}-external-gitcommit-{info["ref"]}.tgz',
+        )
+    )
```

Comment:
```suggestion
    download_path = download_dir.join_within_root(
        info["host"],  # host
        info["namespace"],
        info["repo"],
        f'{info["repo"]}-external-gitcommit-{info["ref"]}.tgz',
    )
```
By creating a new instance of RootedPath, you bypass the security that join_within_root would normally enforce

`Path(*parts)` is an unsafe join, then you take the resulting path - wherever it may lead - and make a new RootedPath out of it

Reply:
Got it. Fixed